File: archive-pre-google/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
1 2 3 4 5 n
prev = null
for node not null
    store node's next in temp
    node sets next to prev
    prev=node
    set node to temp
return prev
n<1<2<3 n
        h
      p
'''
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev=None
        while head:
            nextnode=head.next
            head.next=prev
            prev=head
            head=nextnode
        return prev

File: lc/reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/*
x   1 > 2 > 3 > 4 > 5 > x
                        t    
                    c 
                p   
x < 1 < 2 < 3 < 4 < 5 > x
loop on as long as c is not nil
t holds c's next
current next points to prev
next it: pre moves to c, c moves to t
. > 1 > 2 > x
n < 1 < 2 > x
        p
            c
            t    

*/
func reverseList(head *ListNode) *ListNode {
    var p *ListNode
    c := head
    for ; c != nil; {
        t := c.Next
        c.Next = p
        p, c = c, t
    }
    return p
}